Improves information about container and scopes (#597)
spiral/guide: docs/en/framework/scopes.md

# Framework — IoC Scopes

An essential aspect of developing long-living applications is proper context management. In demonized applications,
you are no longer allowed to treat user requests as a global singleton object and store references to its instance in
your services.

Practically it means that you must explicitly request context while processing user input. Spiral offers a simple way to
manage this by using a global IoC (Inversion of Control) container. This acts as a context carrier that allows you to
request specific instances, scoped to a particular context, as if they were global objects.

## How It Works

When processing a user request, the relevant data for that request is placed in a scoped area of the IoC container. This
data is only available for a limited time while processing that request.

> You can use the `Spiral\Core\ScopeInterface->runScope` method to achieve this. The default Spiral

Reviewer: ```suggestion
You can use the `Spiral\Core\ContainerScopeInterface->runScoped` method to achieve this. The default Spiral
```

container, `Spiral\Core\Container`, implements this interface.

**Here is an example**

```php
use Psr\Container\ContainerInterface;

$container->runScoped(
    closure: function (ContainerInterface $container) {
        dump($container->get(UserContext::class);
    },
    bindings: [UserContext::class => $user,],
);
```

> The first argument of the `runScope` method is an array of values that should be bound to the container within the

Reviewer: ```suggestion
The second argument of the `runScope` method is an array of values that should be bound to the container within the
```

callback's scope.

The callback function receives the container as its argument and can use the container to retrieve the values that were

> bound to it within the callback's scope. This can be useful for creating a temporary scope within which certain values

Reviewer: The closure will be autowired. There any parameters can be. The ContainerInterface will resolved as a container of the new scope.

are available.

> In this example, the `runScope` method is creating a temporary scope within which the `UserContext` class is bound to
> the `$user` variable. The callback function can then use the container to retrieve the `UserContext` object instance,
> which is available within the callback's scope.

Reviewer: ```suggestion
In this example, the `runScoped` method is creating a temporary scope within which the `UserContext` class is bound to
the `$user` variable. The callback function can then use the container to retrieve the `UserContext` object instance,
which is available within the callback's scope.
```

run of the scope is using A`notherImplementation::class`.

### Scope Destroying

When a dependency is resolved within a scope, and that scope is destroyed, the finalize method specified by this
attribute will be called. The purpose of the finalize method is to perform any necessary cleanup or finalization actions
before the scope is destroyed. In this case you can use `Finalize` attribute to control finalizing process. The
attribute provides a convenient way to handle resource cleanup and ensure proper destruction of objects when a scope is
no longer needed.

```php
#[Finalize(method: 'finalize')]
class Foo
{
    public bool $finalized = false;

    public function finalize(Logger $logger): void
    {
        $this->finalized = true;
        $logger->log();
    }
}
```

### Naming a Scope

You can optionally give a name to the scope with the `$name` parameter. This can be useful for tracking or managing
specific scopes.

**Restrictions**

These restrictions ensure proper usage and prevent conflicts within the scope hierarchy. Here are the key restrictions:

1. **Named Scope Uniqueness:** Named scopes within the same hierarchy must have unique names. This means that you cannot

>    have multiple scopes with the same name at the same level of the scope hierarchy. Each named scope should have a

Reviewer: > you cannot have multiple scopes with the same name at the same level of the scope hierarchy

Looks wrong. Parallel scopes are allowed (see №4)

   distinct and meaningful name to avoid any ambiguity.

2. **Default Bindings for Named Scopes:** Named scopes can have default bindings associated with them. It's important to

>    note that default bindings do not affect already created container instances, except for the root container. Default

Reviewer: ```suggestion
   note that changing of default bindings does not affect already created container instances with the same scope name, except for the root container. Default
```

Reviewer: I think the restriction is not a restriction. It's a behavior of the container binder.

   bindings provide a convenient way to set up common dependencies for a particular named scope.

3. **Root Scope Name:** The most parent scope in the hierarchy is named the root scope by default. The root scope serves
   as the top-level scope and provides the foundation for all other scopes. It's the starting point for dependency

>    resolution and can have its own set of default bindings.

Reviewer: The meaning is like this:

> The most parent scope has the name `root` by default.

It's also doesn't look like a restriction.

> 4. **Parallel Named Scopes:** Parallel named scopes are allowed within the hierarchy. This means that you can create
>    multiple named scopes at the same level, as long as their names are unique.

Reviewer: ```suggestion
4. **Parallel Named Scopes:** Parallel named scopes are allowed.
```

### Resolving Rules

When a nested scope does not have its own bindings, it will now resolve dependencies from the parent scope. This allows
for more flexible dependency resolution, with dependencies being resolved in the closest parent scope that has the

> required bindings.

Reviewer: Need to fix the second sentence

### Accessing Scoped Values

> You can access the values set in the scope directly from the container or via method injection in your services or
> controllers. However, this should only be done within the IoC scope.

Reviewer: ```suggestion
You can access the values set in the scope directly from the container or via dependency injection in your services or
controllers. However, this should only be done within the IoC scope.
```

Btw is the sentence required?
